cpp/src/io/parquet/writer_impl.cu

```diff
@@ -1407,17 +1453,60 @@ void writer::impl::write(table_view const& table, std::vector<partition_info> co
-  // compression/decompression performance).
-  // If using the default fragment size, scale it up or down depending on the requested page size.
-  if (max_page_fragment_size_ == cudf::io::default_max_page_fragment_size) {
-    max_page_fragment_size_ = (cudf::io::default_max_page_fragment_size * max_page_size_bytes) /
-                              cudf::io::default_max_page_size_bytes;
+  // 5000 is good enough for up to ~200-character strings. Longer strings and deeply nested columns
+  // will start producing fragments larger than the desired page size, so calculate fragment sizes
+  // for each leaf column.  Skip if the fragment size is not the default.
+  bool is_default_fragment_size = max_page_fragment_size_ == -1;
+  if (is_default_fragment_size) { max_page_fragment_size_ = default_max_page_fragment_size; }
+
+  std::vector<size_type> column_frag_size;
+
+  if (table.num_rows() > 0 && is_default_fragment_size) {
+    std::vector<size_t> column_sizes;
+    std::transform(single_streams_table.begin(),
+                   single_streams_table.end(),
+                   std::back_inserter(column_sizes),
+                   [this](auto const& column) { return column_size(column, stream); });
+
+    // adjust global fragment size if a single fragment will overrun a rowgroup
+    auto const table_size  = std::reduce(column_sizes.begin(), column_sizes.end());
+    auto const avg_row_len = util::div_rounding_up_safe<size_t>(table_size, table.num_rows());
+    if (avg_row_len > 0) {
+      auto const rg_frag_size = util::div_rounding_up_safe(max_row_group_size, avg_row_len);
+      max_page_fragment_size_ = std::min<size_type>(rg_frag_size, max_page_fragment_size_);
+    }
+
+    // dividing page size by average row length will tend to overshoot the desired
+    // page size when there's high variability in the row lengths. instead, shoot
+    // for multiple fragments per page to smooth things out. using 2 was too
+    // unbalanced in final page sizes, so using 4 which seems to be a good
+    // compromise at smoothing things out without getting fragment sizes too small.
+    auto frag_size_fn = [&](auto const& col, size_type col_size) {
+      const int target_frags_per_page = is_fixed_width(col.type()) ? 1 : 4;
+      auto const avg_len =
+        target_frags_per_page * util::div_rounding_up_safe<size_type>(col_size, table.num_rows());
+      if (avg_len > 0) {
+        auto const frag_size = util::div_rounding_up_safe<size_type>(max_page_size_bytes, avg_len);
+        return std::min<size_type>(max_page_fragment_size_, frag_size);
+      } else {
+        return max_page_fragment_size_;
+      }
+    };
+
+    std::transform(single_streams_table.begin(),
+                   single_streams_table.end(),
+                   column_sizes.begin(),
+                   std::back_inserter(column_frag_size),
+                   frag_size_fn);
   }
 
+  bool const has_non_uniform_fragments =
```

Comment:
how inefficient would it be to assume that we always have non uniform fragments?

Reply:
Let me think on this.  Now that the frag stats are computed after the chunks, I think we can get away with a single version of `gather_fragment_stats`.  Then it's just the fragments to deal with. I'll see if there's a better way to deal with that.

Reply:
I reworked this to detect the need to resize the fragments in a different way.  Was able to get rid of one of the gather statistics paths as a bonus.  Thanks for the nudge @vuule !